Dark theme

AndyScherzinger opened the pull request:
Resolves #2188 
improvement for #3107 (AMOLED, needs a black variant, not a dark grey to improve AMOLED)

dark theme implementation has been done by @dan0xii 🎉 

**TODO**
* [x] ~~update to material 1.1.0 ([when released](https://mvnrepository.com/artifact/com.google.android.material/material?repo=google))~~ #4749
* [x] Login screen has messed up colours on the server address box.
* [ ] "Sharing" search box has black-on-grey text in dark mode.
* [ ] Rename dialog doesn't match styling of other edittext boxes.
* [x] Toggle button background colour is difficult to see.
I've lightened the track up in dark mode. Example:
![image](https://user-images.githubusercontent.com/1926270/66062583-95c31f00-e539-11e9-9e20-c7002ee0ae6a.png)
* [x] Share link expiration date picker button text is white-on-white or black-on-grey.
(Fix by dan0xii although the colours could be tweaked).
* [x] Issue in settings where theme is not applied immediately (possibly just API 22 and upwards).
(Possibly fixed. See https://github.com/nextcloud/android/pull/3459#issuecomment-534527849).
* [x] (TBC - dialogs have grey backgrounds) dialogs have a black background - to be discussed. A black dialog on a black activity background can't be distinguished anymore. Maybe we should go for a dark grey tone for the dialogs? (no sure though)
* [x] Preferences text on settings screen stays black on night theme but should be white instead


Note: Floating action button is black-on-primary colour on API 19.

tobiasKaminsky commented:
I have external links on my productive instance, and it seems that they are wrongly colored:
![image](https://user-images.githubusercontent.com/5836855/51173662-30010b00-18b6-11e9-948f-d295cd405494.png)

AndyScherzinger commented:
> I have external links on my productive instance, and it seems that they are wrongly colored:

@tobiasKaminsky yes I also saw that, other than that the PR needs a lot of testing since with the move to AndroidX _and_ supporting day/night via theming I had to abandon the Bridge base themes from the material components, so the theme crashes for legacy AppCompat UI elements (which we shouldn't have in placce anymore, but still), also I can't say if we need to upgrade dialogs, since material components 1.1.0 will introduce new dialog support classes... but well, this is Android dev life...

dan0xii commented:
I've updated the dark_theme branch (hope that works for you) with my latest changes since the move to AndroidX.
TODO are:
Actionbar popup menu on the file list activity.
Popup menus for each file list item
Preference summary text

AndyScherzinger commented:
@dan0xii my AndroidX changes can be found here: https://github.com/nextcloud/android/commit/1e2961e30679278343c4743092ff7b5c48391730 (luckily I still had the local branch and pushed it back to a separate branch now.

AndyScherzinger commented:
@dan0xii no worries, I put it on a separate branch since I don't knwo which of my changes wold still be relevant. I though going for the day/night themes might make sense so besides a light/dark theme we could actually aim for light, dark, day/night and amoled-black for the theme choice while light would be day, and dark would be night.

dan0xii commented:
I _think_ I've incorporated your changes from the dark_themeAS branch. I couldn't for the life of me get cherry-pick to work so I did it manually. The switch back to DayNight in the alpha version of material components has solved a few of the issues.

AndyScherzinger commented:
Thanks for the effort @dan0xii - would you be fine if we (both) just work on this branch, on this repo? That should make it easier for the both of us (while we can't to force-push and rebases anymore ;)

dan0xii commented:
Yes, definitely.

tobiasKaminsky commented:
@AndyScherzinger is dark_themeAS branch still needed?

AndyScherzinger commented:
Yes, so I can do the manual duff if everything needed has been transfered to this branch. After that we can savely delete it.

AndyScherzinger commented:
Open issues I found so far:

* [ ] Auto Upload Dialog Fragment (click on the 3-dot menu on one of the folder-items) has a white background
* [ ] Menu items' text is black (except for the external links), should be white
* [ ] Cancel button text is primary colored and so is the button itself (cancel button should just have a border colored primary since it is a secondary action button) - see "upload file" via bottom menu / FAB
* [ ] dialogs have a black background - to be discussed. A black dialog on a bacl activity backgroudn can't be distinguished anymore. Maybe we should go for a dark grey tone for the dialogs? (no sure though)
* [ ] drawer menu icons on non-checked state on the night/dark theme should be white

AndyScherzinger commented:
❗️ rebased, fixed the issue with the secodary action button theming

dan0xii commented:
You've ticked a couple off my list. On my TODO:
* Highlighted text uses the NC primary blue not the server-set theme colour (Search bar, rename dialog).
* Cursor colour appears primary blue throughout not the server-set theme colour (similar to https://github.com/nextcloud/android/commit/9bc802dddf2f7c430f4ee9ea22dd643df56a5f77)
* File list share icons look strange
* Passcode screen uses primary blue for the "Cancel" button and cursor colour.

AndyScherzinger commented:
@dan0xii I deleted our TODO list comments and added all of them to the main issue's descsription, so we have all things that still need to be taken care of to be found in one central place :)

AndyScherzinger commented:
I can have a look at the share icon thing btw.

AndyScherzinger commented:
✔️ fixed the drawer menu items @dan0xii

dan0xii commented:
@AndyScherzinger 

I've committed what I have on the DatePickerDialog because I'm not getting much further. Buttons need to be fixed and the overall colouring (whether we use the primary colour for the highlighted date or not) needs to be corrected.

AndyScherzinger commented:
> Buttons need to be fixed and the overall colouring (whether we use the primary colour for the highlighted date or not) needs to be corrected.

@dan0xii okay, we'll likely have to go for a "gray" themeing with day/night variants since we can't dynamically theme the picker based on the server-side theme colors.

dan0xii commented:
> @dan0xii okay, we'll likely have to go for a "gray" themeing with day/night variants since we can't dynamically theme the picker based on the server-side theme colors.

Do we not have the option of themeing the current date and selected date parts of the dialog with the server colour (ignoring whether it's day or night) the same way we do for the drawer list items?

dan0xii commented:
> Passcode screen uses primary blue for the "Cancel" button and cursor colour.

"Cancel" button now reflects primary colour.

> backup contacts: primary action buttons are grey instead of primary colored

This is done.

AndyScherzinger commented:
Nice work 👍
If we find a way to properly theme the date picker that would be awesome. In the past we didn't find a way :(

AndyScherzinger commented:
> The problem is solved however I'm assuming you use setColorFilter over setBackgroundColor for a good reason?

Nope your change with setting the color directly is fine with me. This is rather due to an older implementation before using `MaterialButton`. 👍 

Btw I did a rebase since the branch had conflicts with latest master.

tobiasKaminsky commented:
Looks really promising!
Some things I found:
![2019-02-06-153443](https://user-images.githubusercontent.com/5836855/52348689-39991100-2a25-11e9-86a4-e877daa9d7ca.png)
![2019-02-06-153448](https://user-images.githubusercontent.com/5836855/52348688-39991100-2a25-11e9-81fe-3436820836b7.png)

Here the background of the toggle buttons are hard to see.
![2019-02-06-153640](https://user-images.githubusercontent.com/5836855/52348686-39991100-2a25-11e9-9c7c-caaaab704bbb.png)

Upon enabling nothing happens in the settings screen, I have to leave it to get the new mode.

dan0xii commented:
I've pushed the change to FolderPickerDialog and I've ticked it off the todo list. I've also updated your todo list with our progress and added in the problems @tobiasKaminsky found.

AndyScherzinger commented:
Sweet, I am sick this week, so not developing anything this week so feel free to rebase whenever you want to do so. :)

AndyScherzinger commented:
> The problem is solved however I'm assuming you use setColorFilter over setBackgroundColor for a good reason?

@dan0xii Just had this issues in another PR where the method didn't wortrk and the buttons stayed blue. To test this you'll need to have a custon color set on the server's theming to eg red or green or whatever and then see if the buttons will actually change their coloring. With the lib'S version 1.0.0 they seem to not change the color except when explicitely tinted :(

dan0xii commented:
Hm, I see what you mean. I get mixed results for ex.: In FolderDialogPicker the button on the "Local folder" dialog is styled whereas the "Remote folder" is not.

I'll take a look and see if I can fix it in master.

I rebase'd and squashed a couple of my commits.

dan0xii commented:
https://github.com/dan0xii/android/blob/cfbd6b567d792f743a0a7f0d19568ef7083eee18/src/main/java/com/owncloud/android/ui/activity/FolderPickerActivity.java#L93-L94

https://github.com/dan0xii/android/blob/cfbd6b567d792f743a0a7f0d19568ef7083eee18/src/main/java/com/owncloud/android/ui/activity/FolderPickerActivity.java#L414-L428

That appears to work although I've done limited testing and only on 7.1.1. It appears backgroundTint is the way to set background colour for material buttons ([ref](https://github.com/material-components/material-components-android/blob/master/docs/components/MaterialButton.md#Attributes)).

tobiasKaminsky commented:
> [dan0xii/android:src/main/java/com/owncloud/android/ui/activity/FolderPickerActivity.java@`cfbd6b5`#L414-L428](https://github.com/dan0xii/android/blob/cfbd6b567d792f743a0a7f0d19568ef7083eee18/src/main/java/com/owncloud/android/ui/activity/FolderPickerActivity.java#L414-L428)

This is only available for API21+, but we support 14+, so this is unfortunately not working :/
I guess we have to go with the getBackground() approach.

dan0xii commented:
The method appears to work fine on API 15 for me with MaterialButton. Where do you see the requirement for 21+?

tobiasKaminsky commented:
> The method appears to work fine on API 15 for me with MaterialButton. Where do you see the requirement for 21+?

Android Studio is saying this.

![image](https://user-images.githubusercontent.com/5836855/53082112-77556980-34fc-11e9-9f62-b4e71c103d09.png)

mCancel is from type "Button".

tobiasKaminsky commented:
Ah. Nevermind, MaterialButton is indeed working.

AndyScherzinger commented:
@dan0xii would you be able to do a rebase of the branch? I end up having a a large amount of conflicts when trying to rebase :(

dan0xii commented:
> dialogs have a black background - to be discussed. A black dialog on a bacl activity backgroudn can't be distinguished anymore. Maybe we should go for a dark grey tone for the dialogs? (no sure though)

Where do you see this, @AndyScherzinger ? The dialogs have a grey background so I might have changed something since this comment.

Rebase incoming.

AndyScherzinger commented:
Sweet! Don't know about the dialogs anymore,so let's check again at the end.

tobiasKaminsky commented:
![2019-04-02-123833](https://user-images.githubusercontent.com/5836855/55399877-3eb99e80-5544-11e9-9310-70105183e982.png)

I am unsure how this can be solved, but on the left there are black icons (share, lock, etc).

AndyScherzinger commented:
Hmm, there are maybe two ways to handle this
* circular white backround image for the icons
* detect icon color and in case it is black or dark grey and theme is set to dark tint icon white or light grey. Not sure how to do this though, especially for SVGs

tobiasKaminsky commented:
> * detect icon color and in case it is black or dark grey and theme is set to dark tint icon white or light grey. Not sure how to do this though, especially for SVGs

auto tinting icons we cannot control (in theory any app can register their own activity) is something that asks for problems and then it will look strange.



> * circular white backround image for the icons

Nice idea, but will this look properly?

AndyScherzinger commented:
> Nice idea, but will this look properly?

don't know actually, it would kind of behave like launcher icons not compatble with Android 8 with a device having a dark background (Android adding a circular white background).
So it might work I guess.

tobiasKaminsky commented:
> > Nice idea, but will this look properly?
> 
> don't know actually, it would kind of behave like launcher icons not compatble with Android 8 with a device having a dark background (Android adding a circular white background).
> So it might work I guess.

:+1: then we should give it a try

AndyScherzinger commented:
@tobiasKaminsky please review/test and merge if fine with the actual state. As discussed on irc we like to get this in and move on and improve from there. I fixed the most critical things I found from the list with the latest commits. So this should be quite okay for an initial version als minor issues to be found can then be fixed by anyone with smaller PRs and we get a larger amount of testing time on the dev builds before the next stable release 🎉

AndyScherzinger commented:
@tobiasKaminsky @dan0xii please also see the lint errors. My guess: we need to move the datepicker styling to Android 5+ (v21+) and will either have to live with non-dark dat pickers _or_ will have to manually tint them. Also the buttons on the date pickers do have an issue, they now tint the background not the text color which is strange since the API calls seem right.

tobiasKaminsky commented:
![image](https://user-images.githubusercontent.com/5836855/56963893-52175400-6b5a-11e9-8698-2efc3332138b.png)

Maybe the sharing icon on folder can be white instead of black?

tobiasKaminsky commented:
![2019-04-30-151502](https://user-images.githubusercontent.com/5836855/56964091-c2be7080-6b5a-11e9-8f23-a20c7fb6c8b3.png)

The overflow menu is not really visible.
Is it possible to have a slightly more greyish?

tobiasKaminsky commented:
Settings is still white, even if dark is selected:
- have light theming
- switch to dark
- go to files
- go to settings
- still see white

tobiasKaminsky commented:
![2019-04-30-153330](https://user-images.githubusercontent.com/5836855/56965331-55f8a580-6b5d-11e9-8de7-a7e0bfba6f66.png)

This seems indeed to be wrong :rofl: 
If this cannot be tinted dynamically I would go for "white text on black button" for those dialogs.

tobiasKaminsky commented:
There are also some unused resources, maybe coming from dev, please check.

After this is done, I think we can merge it.

AndyScherzinger commented:
> There are also some unused resources, maybe coming from dev, please check.

@tobiasKaminsky can you provide a hint which one(s) that might be? :)

> The overflow menu is not really visible.
Is it possible to have a slightly more greyish?

This is actually how the dark theme looko like so I'd tend to rather keep it that way.

> Settings is still white, even if dark is selected:

yeah after a while it actually gets themed dark don't know why that takes a while atm, maybe @dan0xii does have an idea? :)

> ![2019-04-30-153330](https://user-images.githubusercontent.com/5836855/56965331-55f8a580-6b5d-11e9-8de7-a7e0bfba6f66.png)
If this cannot be tinted dynamically I would go for "white text on black button" for those dialogs.

This is a real challange I guess, haven't found a way to fix this, maybe @dan0xii has an idea. For some reason I don't find a way to access the buttons (which works for the other date picker dialog, not being wrappen in a fragment though).

dan0xii commented:
> > There are also some unused resources, maybe coming from dev, please check.
> 
> @tobiasKaminsky can you provide a hint which one(s) that might be? :)
> 
> > The overflow menu is not really visible.
> > Is it possible to have a slightly more greyish?
> 
> This is actually how the dark theme looko like so I'd tend to rather keep it that way.
> 
> > Settings is still white, even if dark is selected:
> 
> yeah after a while it actually gets themed dark don't know why that takes a while atm, maybe @dan0xii does have an idea? :)
> 
> > ![2019-04-30-153330](https://user-images.githubusercontent.com/5836855/56965331-55f8a580-6b5d-11e9-8de7-a7e0bfba6f66.png)
> > If this cannot be tinted dynamically I would go for "white text on black button" for those dialogs.
> 
> This is a real challange I guess, haven't found a way to fix this, maybe @dan0xii has an idea. For some reason I don't find a way to access the buttons (which works for the other date picker dialog, not being wrappen in a fragment though).


DatePickerDialog was fixed on API 23. Let me check tonight if I've got changes I need to push.

![Screenshot_1556711881](https://user-images.githubusercontent.com/1926270/57016198-c4e30680-6c10-11e9-98a8-2431c17a459d.png)

AndyScherzinger commented:
Thanks @dan0xii beware though that it is the picker for the link expiration date not for the contacts backup. That one is fixed with my latest Commits :)

dan0xii commented:
Fair play. The link expiration picker looks fine on my side as well. Will double-check tonight and sync up with your changes.

tobiasKaminsky commented:
> > There are also some unused resources, maybe coming from dev, please check.
> 
> @tobiasKaminsky can you provide a hint which one(s) that might be? :)

This is found in lint: https://www.kaminsky.me/nc-dev/android-lint/8936.html#UnusedResources



> > The overflow menu is not really visible.
> > Is it possible to have a slightly more greyish?
> 
> This is actually how the dark theme looko like so I'd tend to rather keep it that way.
> 
> >

Okay, did not know this, then we can stick to it.

AndyScherzinger commented:
Okay, then I'd say let's first wait for @dan0xii's commit/push and then check what is left to do before merging.

dan0xii commented:
I've pushed my changes and updated Andy's TODO. I was having trouble merging my changes so I've pushed what I have so that at least other people can contribute. I'm still working on the issues above and on getting the lint score down.

AndyScherzinger commented:
@dan0xii I'll give it a spin today. Can you also do a rebase? (I can do it too) since the master already upgraded the spotbugs library so the build checks will fail due to that. I am fine either way but one of us will then have to clone the repo again.

AndyScherzinger commented:
I did some more minor fixes while I found two (new) issues that need fixing:

* tinting of the menu-checkboxes of the share items doesn't work (anymore... maybe due to a theme change by me?)
<img width="540" alt="device-2019-05-19-183718" src="https://user-images.githubusercontent.com/1315170/57985230-3c17f780-7a65-11e9-968b-777d30f42636.png">

App Crashes after seeting the theme and navigating back
* open the app
* go to the settings screen
* toggle theme
* hit the back arrow in the toolbar
* app terminates

```
2019-05-19 18:39:22.777 29332-29332/com.nextcloud.client E/AndroidRuntime: FATAL EXCEPTION: main
    Process: com.nextcloud.client, PID: 29332
    java.lang.RuntimeException: Unable to start activity ComponentInfo{com.nextcloud.client/com.owncloud.android.ui.activity.FileDisplayActivity}: java.lang.IllegalStateException: This Activity already has an action bar supplied by the window decor. Do not request Window.FEATURE_SUPPORT_ACTION_BAR and set windowActionBar to false in your theme to use a Toolbar instead.
        at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2951)
        at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:3086)
        at android.app.ActivityThread.handleRelaunchActivityInner(ActivityThread.java:4823)
        at android.app.ActivityThread.handleRelaunchActivity(ActivityThread.java:4732)
        at android.app.servertransaction.ActivityRelaunchItem.execute(ActivityRelaunchItem.java:69)
        at android.app.servertransaction.TransactionExecutor.executeCallbacks(TransactionExecutor.java:108)
        at android.app.servertransaction.TransactionExecutor.execute(TransactionExecutor.java:68)
        at android.app.ClientTransactionHandler.executeTransaction(ClientTransactionHandler.java:55)
        at android.app.ActivityThread.handleRelaunchActivityLocally(ActivityThread.java:4782)
        at android.app.ActivityThread.access$3200(ActivityThread.java:200)
        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1826)
        at android.os.Handler.dispatchMessage(Handler.java:106)
        at android.os.Looper.loop(Looper.java:193)
        at android.app.ActivityThread.main(ActivityThread.java:6718)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:493)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:858)
     Caused by: java.lang.IllegalStateException: This Activity already has an action bar supplied by the window decor. Do not request Window.FEATURE_SUPPORT_ACTION_BAR and set windowActionBar to false in your theme to use a Toolbar instead.
        at androidx.appcompat.app.AppCompatDelegateImpl.setSupportActionBar(AppCompatDelegateImpl.java:345)
        at androidx.appcompat.app.AppCompatActivity.setSupportActionBar(AppCompatActivity.java:130)
        at com.owncloud.android.ui.activity.ToolbarActivity.setupToolbar(ToolbarActivity.java:69)
        at com.owncloud.android.ui.activity.ToolbarActivity.setupToolbar(ToolbarActivity.java:99)
        at com.owncloud.android.ui.activity.FileDisplayActivity.onCreate(FileDisplayActivity.java:255)
        at android.app.Activity.performCreate(Activity.java:7144)
        at android.app.Activity.performCreate(Activity.java:7135)
        at android.app.Instrumentation.callActivityOnCreate(Instrumentation.java:1271)
        at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2931)
        at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:3086) 
        at android.app.ActivityThread.handleRelaunchActivityInner(ActivityThread.java:4823) 
        at android.app.ActivityThread.handleRelaunchActivity(ActivityThread.java:4732) 
        at android.app.servertransaction.ActivityRelaunchItem.execute(ActivityRelaunchItem.java:69) 
        at android.app.servertransaction.TransactionExecutor.executeCallbacks(TransactionExecutor.java:108) 
        at android.app.servertransaction.TransactionExecutor.execute(TransactionExecutor.java:68) 
        at android.app.ClientTransactionHandler.executeTransaction(ClientTransactionHandler.java:55) 
        at android.app.ActivityThread.handleRelaunchActivityLocally(ActivityThread.java:4782) 
        at android.app.ActivityThread.access$3200(ActivityThread.java:200) 
        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1826) 
        at android.os.Handler.dispatchMessage(Handler.java:106) 
        at android.os.Looper.loop(Looper.java:193) 
        at android.app.ActivityThread.main(ActivityThread.java:6718) 
        at java.lang.reflect.Method.invoke(Native Method) 
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:493) 
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:858) 
```

@dan0xii fancy taking a look? Beware I rebased the branch ❗️ So you need a fresh checkout and Androd Studio 3.4.1

dan0xii commented:
Will do, let me have a look.

AndyScherzinger commented:
Thanks a million :heart:

Unpublished commented:
<details>
<summary>IllegalStateException</summary>

> App Crashes after seeting the theme and navigating back
> 
>     * open the app
> 
>     * go to the settings screen
> 
>     * toggle theme
> 
>     * hit the back arrow in the toolbar
> 
>     * app terminates
> 
> 
> ```
> 2019-05-19 18:39:22.777 29332-29332/com.nextcloud.client E/AndroidRuntime: FATAL EXCEPTION: main
>     Process: com.nextcloud.client, PID: 29332
>     java.lang.RuntimeException: Unable to start activity ComponentInfo{com.nextcloud.client/com.owncloud.android.ui.activity.FileDisplayActivity}: java.lang.IllegalStateException: This Activity already has an action bar supplied by the window decor. Do not request Window.FEATURE_SUPPORT_ACTION_BAR and set windowActionBar to false in your theme to use a Toolbar instead.
>         at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2951)
>         at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:3086)
>         at android.app.ActivityThread.handleRelaunchActivityInner(ActivityThread.java:4823)
>         at android.app.ActivityThread.handleRelaunchActivity(ActivityThread.java:4732)
>         at android.app.servertransaction.ActivityRelaunchItem.execute(ActivityRelaunchItem.java:69)
>         at android.app.servertransaction.TransactionExecutor.executeCallbacks(TransactionExecutor.java:108)
>         at android.app.servertransaction.TransactionExecutor.execute(TransactionExecutor.java:68)
>         at android.app.ClientTransactionHandler.executeTransaction(ClientTransactionHandler.java:55)
>         at android.app.ActivityThread.handleRelaunchActivityLocally(ActivityThread.java:4782)
>         at android.app.ActivityThread.access$3200(ActivityThread.java:200)
>         at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1826)
>         at android.os.Handler.dispatchMessage(Handler.java:106)
>         at android.os.Looper.loop(Looper.java:193)
>         at android.app.ActivityThread.main(ActivityThread.java:6718)
>         at java.lang.reflect.Method.invoke(Native Method)
>         at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:493)
>         at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:858)
>      Caused by: java.lang.IllegalStateException: This Activity already has an action bar supplied by the window decor. Do not request Window.FEATURE_SUPPORT_ACTION_BAR and set windowActionBar to false in your theme to use a Toolbar instead.
>         at androidx.appcompat.app.AppCompatDelegateImpl.setSupportActionBar(AppCompatDelegateImpl.java:345)
>         at androidx.appcompat.app.AppCompatActivity.setSupportActionBar(AppCompatActivity.java:130)
>         at com.owncloud.android.ui.activity.ToolbarActivity.setupToolbar(ToolbarActivity.java:69)
>         at com.owncloud.android.ui.activity.ToolbarActivity.setupToolbar(ToolbarActivity.java:99)
>         at com.owncloud.android.ui.activity.FileDisplayActivity.onCreate(FileDisplayActivity.java:255)
>         at android.app.Activity.performCreate(Activity.java:7144)
>         at android.app.Activity.performCreate(Activity.java:7135)
>         at android.app.Instrumentation.callActivityOnCreate(Instrumentation.java:1271)
>         at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2931)
>         at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:3086) 
>         at android.app.ActivityThread.handleRelaunchActivityInner(ActivityThread.java:4823) 
>         at android.app.ActivityThread.handleRelaunchActivity(ActivityThread.java:4732) 
>         at android.app.servertransaction.ActivityRelaunchItem.execute(ActivityRelaunchItem.java:69) 
>         at android.app.servertransaction.TransactionExecutor.executeCallbacks(TransactionExecutor.java:108) 
>         at android.app.servertransaction.TransactionExecutor.execute(TransactionExecutor.java:68) 
>         at android.app.ClientTransactionHandler.executeTransaction(ClientTransactionHandler.java:55) 
>         at android.app.ActivityThread.handleRelaunchActivityLocally(ActivityThread.java:4782) 
>         at android.app.ActivityThread.access$3200(ActivityThread.java:200) 
>         at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1826) 
>         at android.os.Handler.dispatchMessage(Handler.java:106) 
>         at android.os.Looper.loop(Looper.java:193) 
>         at android.app.ActivityThread.main(ActivityThread.java:6718) 
>         at java.lang.reflect.Method.invoke(Native Method) 
>         at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:493) 
>         at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:858) 
> ```
> 
> 
</details>
Hey @dan0xii @AndyScherzinger, this should be the same crash as #4026 which now has a fix.

AndyScherzinger commented:
@dan0xii the fix done by @Unpublished resolved the issue with the crash (branch has been rebased by me just seconds ago) 👍

AndyScherzinger commented:
Theming issues I found are with mostly any dialog in preferences and the (pop-up) menu checkboxes.
<img width="540" alt="device-2019-05-19-183718" src="https://user-images.githubusercontent.com/1315170/57985230-3c17f780-7a65-11e9-968b-777d30f42636.png">

preferences
![device-2019-05-20-214316](https://user-images.githubusercontent.com/1315170/58047337-4d860000-7b48-11e9-96f6-c056f9cff497.png)
![device-2019-05-20-214255](https://user-images.githubusercontent.com/1315170/58047339-4e1e9680-7b48-11e9-9773-8a36175690df.png)

It seems that the dialogs are already making use of the new material 2.0 button designs for the primmmary action while for reasons yet unknown they seem to just use funny colors I we need to find out how to overwrite the stilling if this is even possible :/

dan0xii commented:
> @dan0xii the fix done by @Unpublished resolved the issue with the crash (branch has been rebased by me just seconds ago) +1

I find the application hangs when you open Settings and the dark theme is enabled but it's fine when the light theme is set. It seems to go into a loop until you kill the process:

```2019-05-22 13:17:05.906 12881-13834/com.nextcloud.client D/GetRemoteSharesForFileOperation: Got 4 shares
2019-05-22 13:17:05.928 12881-13834/com.nextcloud.client D/FileDataStorageManager: Sending 37 operations to FileContentProvider
2019-05-22 13:17:05.928 12881-13834/com.nextcloud.client D/FileContentProvider: applying batch in provider com.owncloud.android.providers.FileContentProvider@3fb89c9 (temporary: false)
2019-05-22 13:17:05.982 12881-13834/com.nextcloud.client D/FileContentProvider: applied batch in provider com.owncloud.android.providers.FileContentProvider@3fb89c9
2019-05-22 13:17:05.982 12881-13834/com.nextcloud.client D/RefreshFolderOperation: Send broadcast com.owncloud.android.operations.RefreshFolderOperation.EVENT_SINGLE_FOLDER_SHARES_SYNCED
2019-05-22 13:17:05.984 12881-12881/com.nextcloud.client D/FileDisplayActivity: Received broadcast com.owncloud.android.operations.RefreshFolderOperation.EVENT_SINGLE_FOLDER_SHARES_SYNCED
2019-05-22 13:17:06.042 12881-12881/com.nextcloud.client D/FileDisplayActivity: Setting progress visibility to false
2019-05-22 13:17:08.268 12881-12881/com.nextcloud.client V/FileDisplayActivity: onPause() start
2019-05-22 13:17:08.270 12881-12881/com.nextcloud.client D/FileDisplayActivity: onPause() ending
2019-05-22 13:17:08.270 12881-12881/com.nextcloud.client V/FileDisplayActivity: onPause() end
2019-05-22 13:17:08.299 12881-12881/com.nextcloud.client D/SettingsActivity: onCreate(Bundle) starting
2019-05-22 13:17:08.416 12881-12881/com.nextcloud.client D/SettingsActivity: onStart() starting
2019-05-22 13:17:08.418 12881-12881/com.nextcloud.client D/SettingsActivity: onResume() starting
2019-05-22 13:17:08.568 12881-12948/com.nextcloud.client D/EGL_emulation: eglMakeCurrent: 0xb03a18c0: ver 2 0 (tinfo 0xac07bc10)
2019-05-22 13:17:08.651 12881-12948/com.nextcloud.client D/EGL_emulation: eglMakeCurrent: 0xb03a18c0: ver 2 0 (tinfo 0xac07bc10)
2019-05-22 13:17:08.698 12881-12948/com.nextcloud.client D/EGL_emulation: eglMakeCurrent: 0xb03a18c0: ver 2 0 (tinfo 0xac07bc10)
2019-05-22 13:17:09.053 12881-12881/com.nextcloud.client V/FileDisplayActivity: onSaveInstanceState() start
2019-05-22 13:17:09.054 12881-12881/com.nextcloud.client D/FileDisplayActivity: onSaveInstanceState(Bundle) starting
2019-05-22 13:17:09.054 12881-12881/com.nextcloud.client D/ExtendedListFragment: onSaveInstanceState()
2019-05-22 13:17:09.055 12881-12881/com.nextcloud.client V/FileDisplayActivity: onSaveInstanceState() end
2019-05-22 13:17:09.057 12881-12881/com.nextcloud.client D/FileDisplayActivity: onStop() ending
2019-05-22 13:17:14.644 12881-12881/com.nextcloud.client W/System.err: java.lang.NoSuchMethodException: performStopActivity [interface android.os.IBinder, boolean]
2019-05-22 13:17:14.667 12881-12881/com.nextcloud.client W/System.err:     at java.lang.Class.getMethod(Class.java:2068)
2019-05-22 13:17:14.667 12881-12881/com.nextcloud.client W/System.err:     at java.lang.Class.getDeclaredMethod(Class.java:2047)
2019-05-22 13:17:14.667 12881-12881/com.nextcloud.client W/System.err:     at androidx.core.app.ActivityRecreator.getPerformStopActivity2Params(ActivityRecreator.java:304)
2019-05-22 13:17:14.667 12881-12881/com.nextcloud.client W/System.err:     at androidx.core.app.ActivityRecreator.<clinit>(ActivityRecreator.java:94)
2019-05-22 13:17:14.667 12881-12881/com.nextcloud.client W/System.err:     at androidx.core.app.ActivityRecreator.recreate(ActivityRecreator.java:105)
2019-05-22 13:17:14.667 12881-12881/com.nextcloud.client W/System.err:     at androidx.core.app.ActivityCompat.recreate(ActivityCompat.java:580)
2019-05-22 13:17:14.667 12881-12881/com.nextcloud.client W/System.err:     at androidx.appcompat.app.AppCompatDelegateImpl.updateForNightMode(AppCompatDelegateImpl.java:2248)
2019-05-22 13:17:14.667 12881-12881/com.nextcloud.client W/System.err:     at androidx.appcompat.app.AppCompatDelegateImpl.applyDayNight(AppCompatDelegateImpl.java:2123)
2019-05-22 13:17:14.667 12881-12881/com.nextcloud.client W/System.err:     at androidx.appcompat.app.AppCompatDelegateImpl.applyDayNight(AppCompatDelegateImpl.java:2117)
2019-05-22 13:17:14.667 12881-12881/com.nextcloud.client W/System.err:     at com.owncloud.android.ui.activity.SettingsActivity.lambda$setupGeneralCategory$18$SettingsActivity(SettingsActivity.java:706)
2019-05-22 13:17:14.667 12881-12881/com.nextcloud.client W/System.err:     at com.owncloud.android.ui.activity.-$$Lambda$SettingsActivity$7clUhqQRpIRKO9qSP_Irc5AiJHw.onPreferenceChange(Unknown Source:2)
2019-05-22 13:17:14.667 12881-12881/com.nextcloud.client W/System.err:     at android.preference.Preference.callChangeListener(Preference.java:1079)
2019-05-22 13:17:14.668 12881-12881/com.nextcloud.client W/System.err:     at android.preference.TwoStatePreference.onClick(TwoStatePreference.java:65)
2019-05-22 13:17:14.668 12881-12881/com.nextcloud.client W/System.err:     at android.preference.Preference.performClick(Preference.java:1134)
2019-05-22 13:17:14.668 12881-12881/com.nextcloud.client W/System.err:     at android.preference.PreferenceScreen.onItemClick(PreferenceScreen.java:249)
2019-05-22 13:17:14.668 12881-12881/com.nextcloud.client W/System.err:     at android.widget.AdapterView.performItemClick(AdapterView.java:318)
2019-05-22 13:17:14.669 12881-12881/com.nextcloud.client W/System.err:     at android.widget.AbsListView.performItemClick(AbsListView.java:1165)
2019-05-22 13:17:14.669 12881-12881/com.nextcloud.client W/System.err:     at android.widget.AbsListView$PerformClick.run(AbsListView.java:3134)
2019-05-22 13:17:14.669 12881-12881/com.nextcloud.client W/System.err:     at android.widget.AbsListView$3.run(AbsListView.java:4049)
2019-05-22 13:17:14.669 12881-12881/com.nextcloud.client W/System.err:     at android.os.Handler.handleCallback(Handler.java:789)
2019-05-22 13:17:14.669 12881-12881/com.nextcloud.client W/System.err:     at android.os.Handler.dispatchMessage(Handler.java:98)
2019-05-22 13:17:14.669 12881-12881/com.nextcloud.client W/System.err:     at android.os.Looper.loop(Looper.java:164)
2019-05-22 13:17:14.669 12881-12881/com.nextcloud.client W/System.err:     at android.app.ActivityThread.main(ActivityThread.java:6541)
2019-05-22 13:17:14.669 12881-12881/com.nextcloud.client W/System.err:     at java.lang.reflect.Method.invoke(Native Method)
2019-05-22 13:17:14.670 12881-12881/com.nextcloud.client W/System.err:     at com.android.internal.os.Zygote$MethodAndArgsCaller.run(Zygote.java:240)
2019-05-22 13:17:14.670 12881-12881/com.nextcloud.client W/System.err:     at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:767)
2019-05-22 13:17:14.730 12881-12881/com.nextcloud.client D/SettingsActivity: onPause() ending
2019-05-22 13:17:14.730 12881-12881/com.nextcloud.client D/SettingsActivity: onSaveInstanceState(Bundle) starting
2019-05-22 13:17:14.738 12881-12948/com.nextcloud.client D/EGL_emulation: eglMakeCurrent: 0xb03a18c0: ver 2 0 (tinfo 0xac07bc10)
2019-05-22 13:17:14.740 12881-12881/com.nextcloud.client D/SettingsActivity: onStop() ending
2019-05-22 13:17:14.741 12881-12881/com.nextcloud.client D/SettingsActivity: onDestroy() ending
2019-05-22 13:17:14.747 12881-12948/com.nextcloud.client D/OpenGLRenderer: endAllActiveAnimators on 0x8cfd2880 (ListView) with handle 0xac07bf00
2019-05-22 13:17:14.767 12881-12881/com.nextcloud.client D/SettingsActivity: onCreate(Bundle) starting
2019-05-22 13:17:14.845 12881-12881/com.nextcloud.client D/SettingsActivity: onStart() starting
2019-05-22 13:17:14.846 12881-12881/com.nextcloud.client D/SettingsActivity: onResume() starting
2019-05-22 13:17:14.987 12881-12881/com.nextcloud.client D/SettingsActivity: onPause() ending
2019-05-22 13:17:14.988 12881-12881/com.nextcloud.client D/SettingsActivity: onSaveInstanceState(Bundle) starting
2019-05-22 13:17:15.002 12881-12881/com.nextcloud.client D/SettingsActivity: onStop() ending
2019-05-22 13:17:15.003 12881-12881/com.nextcloud.client D/SettingsActivity: onDestroy() ending
2019-05-22 13:17:15.045 12881-12881/com.nextcloud.client D/SettingsActivity: onCreate(Bundle) starting
2019-05-22 13:17:15.101 12881-12881/com.nextcloud.client D/SettingsActivity: onStart() starting
2019-05-22 13:17:15.103 12881-12881/com.nextcloud.client D/SettingsActivity: onResume() starting
2019-05-22 13:17:15.216 12881-12941/com.nextcloud.client D/SharedPreferencesImpl: Time required to fsync /data/user/0/com.nextcloud.client/shared_prefs/com.nextcloud.client_preferences.xml: [<1: 0, <2: 0, <4: 5, <8: 5, <16: 1, <32: 4, <64: 1, <128: 0, <256: 0, <512: 1, <1024: 0, <2048: 0, <4096: 0, <8192: 0, <16384: 0, >=16384: 0]
2019-05-22 13:17:15.248 12881-12881/com.nextcloud.client D/SettingsActivity: onPause() ending
2019-05-22 13:17:15.248 12881-12881/com.nextcloud.client D/SettingsActivity: onSaveInstanceState(Bundle) starting
2019-05-22 13:17:15.260 12881-12881/com.nextcloud.client D/SettingsActivity: onStop() ending
2019-05-22 13:17:15.261 12881-12881/com.nextcloud.client D/SettingsActivity: onDestroy() ending
2019-05-22 13:17:15.309 12881-12881/com.nextcloud.client D/SettingsActivity: onCreate(Bundle) starting
2019-05-22 13:17:15.367 12881-12881/com.nextcloud.client D/SettingsActivity: onStart() starting
2019-05-22 13:17:15.369 12881-12881/com.nextcloud.client D/SettingsActivity: onResume() starting
2019-05-22 13:17:15.704 12881-12881/com.nextcloud.client D/SettingsActivity: onPause() ending
2019-05-22 13:17:15.704 12881-12881/com.nextcloud.client D/SettingsActivity: onSaveInstanceState(Bundle) starting
2019-05-22 13:17:15.707 12881-12881/com.nextcloud.client D/SettingsActivity: onStop() ending
2019-05-22 13:17:15.708 12881-12881/com.nextcloud.client D/SettingsActivity: onDestroy() ending
2019-05-22 13:17:15.746 12881-12881/com.nextcloud.client D/SettingsActivity: onCreate(Bundle) starting
2019-05-22 13:17:15.798 12881-12941/com.nextcloud.client D/SharedPreferencesImpl: Time required to fsync /data/user/0/com.nextcloud.client/shared_prefs/com.nextcloud.client_preferences.xml: [<1: 0, <2: 0, <4: 5, <8: 5, <16: 2, <32: 4, <64: 1, <128: 0, <256: 0, <512: 2, <1024: 0, <2048: 0, <4096: 0, <8192: 0, <16384: 0, >=16384: 0]
2019-05-22 13:17:15.819 12881-12881/com.nextcloud.client D/SettingsActivity: onStart() starting
2019-05-22 13:17:15.824 12881-12881/com.nextcloud.client D/SettingsActivity: onResume() starting
2019-05-22 13:17:16.000 12881-12881/com.nextcloud.client D/SettingsActivity: onPause() ending
2019-05-22 13:17:16.000 12881-12881/com.nextcloud.client D/SettingsActivity: onSaveInstanceState(Bundle) starting
2019-05-22 13:17:16.002 12881-12881/com.nextcloud.client D/SettingsActivity: onStop() ending
2019-05-22 13:17:16.002 12881-12881/com.nextcloud.client D/SettingsActivity: onDestroy() ending
2019-05-22 13:17:16.053 12881-12881/com.nextcloud.client D/SettingsActivity: onCreate(Bundle) starting
`

Unpublished commented:
Looks like the same issue as on news-android see https://github.com/nextcloud/news-android/pull/743#discussion_r270642999.

I think @David-Development fixed the restart loop on news-android by migrating the `SettingsActivity` from `PreferenceActivity` to `AppCompatActivity`+ `PreferenceFragment` [here](https://github.com/nextcloud/news-android/commit/26c90c228cd138856371ef042839e10a84bf8ef2#diff-a5c9346e5f123fb836be66fe3efb6500) and [here](https://github.com/nextcloud/news-android/commit/8af4210fc819b59491d9664b4f70a2ae2d81ba24).

AndyScherzinger commented:
@dan0xii I did some more minor tweakings, also I realized that the text coloring is wrong (inverted) on the settings screen, could you maybe take a look at it? I am clueless at the moment as to why that is :(

tobiasKaminsky commented:
https://proandroiddev.com/making-android-lint-theme-aware-6285737b13bc

Just read it quickly, not sure if this helps :-)

AndyScherzinger commented:
@tobiasKaminsky any idea why the build fails completely?

```
FAILURE: Build failed with an exception.
--
* What went wrong:
--
Gradle build daemon disappeared unexpectedly (it may have been killed or may have crashed)
```

While I don't see any gradle/demon issues in the build logs :/

tobiasKaminsky commented:
When compiling locally, I do get 5 errors from EtmPreferencesFragment.kt.
I do not get them when compiling on master…

AndyScherzinger commented:
@tobiasKaminsky I changed that with the latest commit now while the builds fail completely since post-06.06.2019 so it has nothing to do with this but is some other, general issue :/

tobiasKaminsky commented:
> @tobiasKaminsky I changed that with the latest commit now while the builds fail completely since post-06.06.2019 so it has nothing to do with this but is some other, general issue :/

But only on this branch, or? 
I'll have a look.

tobiasKaminsky commented:
I can locally build & run dark theme without any problems.
@ezaquarii pointed out that it might be a memory problem as drone worker do not have that much memory.
I am unsure, as it fails nearly immediately…

AndyScherzinger commented:
I'd have to check other PRs but think this might happen on other branches too...

tobiasKaminsky commented:
> I'd have to check other PRs but think this might happen on other branches too...

If this is then a more generic problem, we should open up a new ticket and link to those PRs/branches/failing drone so we maybe see the common problem.

AndyScherzinger commented:
Just cehcked and couldn't find any other branch failing for the same reasons (they all have compile errors in contrast to dark_theme)

dan0xii commented:
> * [ ] Issue in settings where theme is not applied immediately (possibly just API 22 and upwards).

Attempted to make this slicker by having BaseActivity and ThemedPreferenceActivity detect changes to the theme and act appropriately. It's quite slow when moving from the SettingsActivity back to the file list and I'm not sure at this point whether I'm interfering with any other processes--could do with some testing.

AndyScherzinger commented:
> It's quite slow when moving from the SettingsActivity back to the file list and I'm not sure at this point whether I'm interfering with any other processes--could do with some testing.

I experience this on master with most navigation paths through the main navigation. My guess is that this isn't related to this PR but rather a general issue already present on master (so I think we should ignore this for the dark theme development). cc @tobiasKaminsky for feedback ❤️ 

> Attempted to make this slicker by having BaseActivity and ThemedPreferenceActivity detect changes to the theme and act appropriately. 

Sweet, will give it a spin 🎢

tobiasKaminsky commented:
> I experience this on master with most navigation paths through the main navigation. My guess is that this isn't related to this PR but rather a general issue already present on master (so I think we should ignore this for the dark theme development). cc @tobiasKaminsky for feedback heart

Can you open up a new issue for this, with steps how to reproduce? :)

dan0xii commented on src/main/java/com/owncloud/android/ui/activity/BaseActivity.java:
```diff
@@ -52,12 +55,53 @@
+    /**
+     * Tracks whether the activity should be recreate()'d after a theme change
+     */
+    private boolean mThemeChangePending = false;
```
Removed

tobiasKaminsky commented:
There is now a beta01 available.
Hopefully this means that 1.1.0 stable will be soon there.

AndyScherzinger commented:
Sweet! While that means we still have quite some work to do (looking at themeing the dialog boxes and their buttons)

AndyScherzinger commented:
❗️ **REBASED** ❗️
cc @dan0xii

AndyScherzinger commented:
Looking good so far, minor issues I found:

* styling for some dialogs, especially for th sharing parts within the details sccreen
* dialogs on setting page aren't themed accordingly

![device-2019-10-07-172130](https://user-images.githubusercontent.com/1315170/66325279-427d1200-e927-11e9-8140-890f600c477a.png)
![device-2019-10-07-171909](https://user-images.githubusercontent.com/1315170/66325280-427d1200-e927-11e9-8876-c506acd7cedd.png)
![device-2019-10-07-171933](https://user-images.githubusercontent.com/1315170/66325282-4315a880-e927-11e9-85ea-2931515642da.png)
![device-2019-10-07-171958](https://user-images.githubusercontent.com/1315170/66325283-4315a880-e927-11e9-9cf2-e3d867c386f5.png)
![device-2019-10-07-172043](https://user-images.githubusercontent.com/1315170/66325284-4315a880-e927-11e9-8b79-067396d79d43.png)

dan0xii commented:
Hoping to have a finished version v soon !

szaimen commented:
One Small Issue I encontered:
When upload isn't allowed, the plus button should be grey like in the current version of this app.
Dark theme app:
![photo5249300370973371460](https://user-images.githubusercontent.com/42591237/66340909-83d0ea00-e946-11e9-9e08-940e1a9cb4c4.jpg)
Current 3.8.0:
![photo5251307379126021276](https://user-images.githubusercontent.com/42591237/66340954-94816000-e946-11e9-94a1-ebb892c959f7.jpg)

AndyScherzinger commented:
@dan0xii can you push your latest changes? (Just asking since to made some changes to the issue's text so I guess you made some, more progress :)

dan0xii commented:
My latest changes are already here. If I edited the todo list after my last commit it was probably on the back of testing I had been doing.

splitt3r commented:
Another small issue:

![image](https://user-images.githubusercontent.com/1869584/67662254-4cf95d00-f963-11e9-85e4-b3e979e71864.png)

All in all in already looks pretty awesome ❗️ Thanks for the great work 👍

tobiasKaminsky commented:
I would like to merge this, even with current problems soon, so that we can enhance this step by stop.
Today I release 3.9.0 RC1, so if we merge it afterwards we have enough time to polish it for 3.10, and get a lot of feedback due to dev channel.

I will test this later, if does not create problems on the "light theme", then update & merge it, after your ok @dan0xii @AndyScherzinger

AndyScherzinger commented:
@tobiasKaminsky I'll do a rebase later today and beware that the colloring issues will be present for the light theme also! So there are things that _must_ be fixed before 3.10 in case this gets merged to mmaster _now_

tobiasKaminsky commented:
>  beware that the coloring issues will be present for the light theme also

Damn, I assumed that this this only for dark.
Then I'll have a look into the coloring issue soon.

AndyScherzinger commented:
> Damn, I assumed that this this only for dark.
Then I'll have a look into the coloring issue soon.

@tobiasKaminsky I am currently taking a look at some aspedcts and will post an update. The reason is that this PR moved to Material classes and that means that some things need to be themed differently (code-wise). Most stuff should already be working now.

I'll keep you posted 👍

dan0xii commented:
I've just pushed a few changes for styling. Corrects a couple of issues pointed out earlier.

AndyScherzinger commented:
@splitt3r thanks for testing, your issues should now be fixed with the latesdt commits 🚀 

@tobiasKaminsky @dan0xii I'd be fine with merging this to master and start with small(er) fixes for papercuts

dan0xii commented:
> @tobiasKaminsky @dan0xii I'd be fine with merging this to master and start with small(er) fixes for papercuts

Agreed. I'll just continue to try and tidy up the theme to get it to _look_ right and then I'll go back and try and make the code slicker.

tobiasKaminsky commented on src/main/java/com/nextcloud/client/onboarding/FirstRunActivity.java:
```diff
@@ -95,7 +94,7 @@ protected void onCreate(Bundle savedInstanceState) {
 
         Button providerButton = findViewById(R.id.signup);
-        providerButton.setBackgroundColor(getResources().getColor(R.color.primary_dark));
+        providerButton.setBackgroundColor(getResources().getColor(R.color.primary));
```
This might lead to problems  with branding, please revert.

tobiasKaminsky commented on src/main/java/com/owncloud/android/datamodel/ThumbnailsCacheManager.java:
```diff
@@ -1173,7 +1173,7 @@ private static Bitmap handlePNG(Bitmap bitmap, int pxW, int pxH) {
         Canvas c = new Canvas(resultBitmap);
 
-        c.drawColor(MainApp.getAppContext().getResources().getColor(R.color.background_color));
+        c.drawColor(MainApp.getAppContext().getResources().getColor(R.color.bg_default));
```
If we change this on demand to dark/light value, this will be permanent in cache and thus might be wrong.
Doing this dynamic would be a way, but for now I think we should just use light background for PNG as always.
(I'll create a ticket to investigate this later).

tobiasKaminsky commented on src/main/java/com/owncloud/android/ui/ThemeableSwitchPreference.java:
```diff
@@ -76,12 +77,15 @@ private void findSwitch(ViewGroup viewGroup) {
+                            new int[][]{new int[]{android.R.attr.state_checked},
+                                new int[]{}},
+                            new int[]{trackColor, trackColorUnchecked});
+//                            new int[]{trackColor, Color.parseColor("#4D000000")});
```
remove commented line

tobiasKaminsky commented on src/main/java/com/owncloud/android/ui/activity/BaseActivity.java:
```diff
@@ -6,25 +6,28 @@
  * Base activity with common behaviour for activities dealing with ownCloud {@link Account}s .
  */
-public abstract class BaseActivity extends AppCompatActivity implements Injectable {
+public abstract class BaseActivity extends AppCompatActivity implements Injectable, SharedPreferences.OnSharedPreferenceChangeListener {
```
Line wrap

tobiasKaminsky commented on src/main/java/com/owncloud/android/ui/activity/BaseActivity.java:
```diff
@@ -42,12 +45,53 @@
+     * Tracks whether the activity should be recreate()'d after a theme change
+     */
+    private boolean mThemeChangePending;
+    private boolean mPaused;
```
no m prefix

tobiasKaminsky commented on src/main/java/com/owncloud/android/ui/activity/BaseActivity.java:
```diff
@@ -42,12 +45,53 @@
+        mPaused = false;
+
+        if(mThemeChangePending) {
+//            getDelegate().applyDayNight();
```
remove commented line

tobiasKaminsky commented on src/main/java/com/owncloud/android/ui/activity/DrawerActivity.java:
```diff
@@ -989,7 +991,12 @@ protected void setDrawerMenuItemChecked(int menuItemId) {
+                        || menuItem.getItemId() == R.id.drawer_menu_account_manage) {
+                        ThemeUtils.tintDrawable(menuItem.getIcon(), ContextCompat.getColor(this, R.color.drawer_menu_icon));
+                    }
+                    menuItem.setTitle(Html.fromHtml("<font color='" + ThemeUtils.colorToHexString(ContextCompat.getColor(this, R.color.textColor)) + "'>" + menuItem.getTitle() + "</font>"));
```
space between if and () (hint: use code formatting)
line wrap

tobiasKaminsky commented on src/main/java/com/owncloud/android/ui/activity/SettingsActivity.java:
```diff
@@ -692,6 +692,17 @@ private void setupGeneralCategory(int accentColor) {
+                            getString(R.string.prefs_value_theme_dark) : getString(R.string.prefs_value_theme_light));
+        themePref.setOnPreferenceChangeListener((preference, newValue) -> {
+            MainApp.setAppTheme((Boolean) newValue);
+//            recreate();
```
remove

tobiasKaminsky commented on src/main/java/com/owncloud/android/ui/activity/ThemedPreferenceActivity.java:
```diff
@@ -0,0 +1,77 @@
+
+import androidx.annotation.Nullable;
+
+public class ThemedPreferenceActivity extends PreferenceActivity implements SharedPreferences.OnSharedPreferenceChangeListener {
```
line wrap

tobiasKaminsky commented on src/main/java/com/owncloud/android/ui/activity/ThemedPreferenceActivity.java:
```diff
@@ -0,0 +1,77 @@
+     * Tracks whether the activity should be recreate()'d after a theme change
+     */
+    private boolean mThemeChangePending;
+    private boolean mPaused;
```
no m prefix

tobiasKaminsky commented on src/main/java/com/owncloud/android/ui/adapter/ActivityListAdapter.java:
```diff
@@ -369,7 +368,8 @@ public void updateDrawState(@NonNull TextPaint ds) {
                 ssb.setSpan(new StyleSpan(android.graphics.Typeface.BOLD), idx1, idx2, 0);
-                ssb.setSpan(new ForegroundColorSpan(Color.BLACK), idx1, idx2, Spannable.SPAN_EXCLUSIVE_EXCLUSIVE);
+                ssb.setSpan(new ForegroundColorSpan(context.getResources().getColor(R.color.textColor)),
+                            idx1, idx2, Spannable.SPAN_EXCLUSIVE_EXCLUSIVE);
```
for such methods with many parameters, we agreed to have a line break after each parameter, so pleas:

```suggestion
                            ssb.setSpan(new ForegroundColorSpan(context.getResources().getColor(R.color.textColor)),
                            idx1, 
                            idx2, 
                            Spannable.SPAN_EXCLUSIVE_EXCLUSIVE);
```

dan0xii replied:
> for such methods with many parameters, we agreed to have a line break after each parameter, so pleas:

Where can I find information on this and general code formatting? I don't see anything in the contribution document but I'll admit I haven't looked through the developer docs.

tobiasKaminsky commented on src/main/java/com/owncloud/android/ui/adapter/NotificationListAdapter.java:
```diff
@@ -87,6 +87,7 @@ public NotificationListAdapter(OwnCloudClient client, NotificationsActivity noti
         this.notificationsList = new ArrayList<>();
         this.client = client;
         this.notificationsActivity = notificationsActivity;
+        foregroundColorSpanBlack = new ForegroundColorSpan(notificationsActivity.getResources().getColor(R.color.textColor));
```
line wrap

tobiasKaminsky commented on src/main/java/com/owncloud/android/ui/fragment/ExtendedListFragment.java:
```diff
@@ -194,6 +194,13 @@ public void onCreateOptionsMenu(Menu menu, MenuInflater inflater) {
+        theTextArea.setHighlightColor(ThemeUtils.primaryAccentColor(getContext()));
+
+//        EditText searchText = searchView.findViewById(R.id.searchView);
+//        searchText.setHighlightColor(ThemeUtils.primaryColor(getContext(), true));
```
Don't we need this?

AndyScherzinger replied:
I think due to Material classes the new code right above takes care of this.

tobiasKaminsky commented on src/main/java/com/owncloud/android/ui/fragment/contactsbackup/ContactsBackupFragment.java:
```diff
@@ -176,19 +176,15 @@ public void onCheckedChanged(CompoundButton buttonView, boolean isChecked) {
         backupNow.setTextColor(fontColor);
 
-        contactsDatePickerBtn.getBackground().setColorFilter(accentColor, PorterDuff.Mode.SRC_ATOP);
+        contactsDatePickerBtn.setBackgroundColor(primaryColor);
```
Is this changed on purpose or by accident?

AndyScherzinger replied:
on purpose this is due to the fact that we are using Material classes (which offer a different/more-direct API, see button above :)

tobiasKaminsky commented on src/main/java/com/owncloud/android/utils/ThemeUtils.java:
```diff
@@ -126,10 +126,12 @@ public static int primaryColor(Context context, boolean replaceWhite) {
     public static int primaryColor(Account account, boolean replaceWhite, Context context) {
-        OCCapability capability = getCapability(account, context);
+        if (context==null) {
+            return 255;
```
Magic number?
Code formatting, please

tobiasKaminsky commented on src/main/java/com/owncloud/android/utils/ThemeUtils.java:
```diff
@@ -507,7 +509,7 @@ public static void themeEditText(Context context, EditText editText, boolean the
         }
 
-        editText.setHintTextColor(color);
+//        editText.setHintTextColor(color);
```
if we do not need it, please remove

tobiasKaminsky commented on src/main/res/drawable-night/round_bgnd.xml:
```diff
@@ -0,0 +1,5 @@
+<?xml version="1.0" encoding="utf-8"?>
```
license header

tobiasKaminsky commented on src/main/res/drawable-night/shared_via_link.xml:
```diff
@@ -0,0 +1,18 @@
+<vector xmlns:android="http://schemas.android.com/apk/res/android"
```
license header

tobiasKaminsky commented on src/main/res/drawable-night/shared_via_users.xml:
```diff
@@ -0,0 +1,12 @@
+<vector xmlns:android="http://schemas.android.com/apk/res/android"
```
license header

tobiasKaminsky commented on src/main/res/layout-land/account_setup.xml:
```diff
@@ -1,225 +1,227 @@
-    </ScrollView>
-
-</LinearLayout>
+<?xml version="1.0" encoding="utf-8"?><!--
```
It this is such a big change, we can also update license header to reflect to Nextcloud app

tobiasKaminsky commented on src/main/res/layout/account_setup.xml:
```diff
@@ -1,225 +1,228 @@
-
-</ScrollView>
+<?xml version="1.0" encoding="utf-8"?><!--
+  ownCloud Android client application
```
Same here

tobiasKaminsky commented on src/main/res/layout/activity_list_item.xml:
```diff
@@ -25,17 +25,20 @@
         android:id="@+id/activity_icon"
-        android:layout_width="@dimen/activity_icon_width"
-        android:layout_height="@dimen/activity_icon_height"
+        android:layout_width="@dimen/notification_icon_width"
```
This is an activity_icon.
If this is the same value as in notification_icon, please duplicate it.
Otherwise we would possibly get nasty side effects, when we change it in one location.

tobiasKaminsky commented on src/main/res/layout/ssl_validator_layout.xml:
```diff
@@ -31,7 +31,7 @@
 		android:paddingBottom="@dimen/standard_padding"
 		android:textAppearance="?android:attr/textAppearanceMedium"
-		android:textColor="@color/black"
+		android:textColor="@color/textColor_inverse"
```
This should be textColor, or? (which is normally black)

tobiasKaminsky commented on src/main/res/layout/toolbar_user_information.xml:
```diff
@@ -91,11 +91,11 @@
                         android:shadowDy="0"
                         android:shadowRadius="2"
-                        android:textColor="@android:color/white"
+                        android:textColor="@color/textColor"
```
TextColor is normally black.
So textColor_inverse, or?

AndyScherzinger replied:
Nice catch, this should actually always be `@android:color/white`

tobiasKaminsky commented on src/main/res/layout/toolbar_user_information.xml:
```diff
@@ -107,11 +107,11 @@
                         android:shadowDy="0"
                         android:shadowRadius="2"
-                        android:textColor="@android:color/white"
+                        android:textColor="@color/textColor"
```
Same here? textColor_inverse?

AndyScherzinger replied:
same here

tobiasKaminsky commented on src/main/res/values-night/colors.xml:
```diff
@@ -0,0 +1,82 @@
+<?xml version="1.0" encoding="utf-8"?>
+<!--
+  ownCloud Android client application
```
New license header, as it is a new file

tobiasKaminsky commented on src/main/res/values-v21/styles.xml:
```diff
@@ -17,6 +17,45 @@
+
+    <style name="DatePickerStyle" parent="">
+        <item name="android:headerBackground">@color/bg_fallback_highlight</item>
+        <!-- TODO for < API21 -->
```
Is this a real todo? Then please remove it here and create a new issue with label "darkTheme".

AndyScherzinger replied:
removed since v16 simply shows a (holo) date picker atm. cc @dan0xii

tobiasKaminsky commented on src/main/res/values/colors.xml:
```diff
@@ -23,14 +23,15 @@
     <color name="black">#000000</color>
-    <color name="white">#FFFFFF</color>
-    <color name="textColor">@color/black</color>
+    <!--<color name="white">#FFFFFF</color>-->
```
remove

tobiasKaminsky commented on src/main/res/values/strings.xml:
```diff
@@ -56,6 +56,11 @@
     <string name="prefs_imprint">Imprint</string>
+    <string name="prefs_value_theme_light">Light</string>
+    <string name="prefs_value_theme_dark">Dark</string>
+    <string name="prefs_key_theme">darkTheme</string>
```
This is a preference internal key, please move it

tobiasKaminsky commented on src/main/res/xml/preferences.xml:
```diff
@@ -1,5 +1,4 @@
-<?xml version="1.0" encoding="utf-8"?>
-<!--
+<?xml version="1.0" encoding="utf-8"?><!--
```
should be 2 lines (though I know that Android Studio likes to put this into one line…)

tobiasKaminsky commented:
Only minor changes due to code review.
@dan0xii @AndyScherzinger this is an amazing work! :tada: :tada: :tada: :tada: 

Once it is merged, I'll create some issues so that we can fix this in smaller steps.

AndyScherzinger commented:
taking care of the review comments right now

AndyScherzinger commented:
@tobiasKaminsky code review changes are ✔️ **DONE**

Feel free to re-review, test-drive and merge :)

AndyScherzinger commented:
Hi everyone,
the initial work has been merged to master an will be paart of tomorrow's dev build. All further findings have been labeled `feature: theming`

dan0xii commented:
So glad to get this merged. Thanks for the help getting it over the line, @AndyScherzinger.

tobiasKaminsky commented:
And this is the link: https://github.com/nextcloud/android/issues?q=is%3Aopen+is%3Aissue+label%3A%22feature%3A+theming%22

Some are bugs, some are enhancements :-)

AndyScherzinger commented:
> So glad to get this merged. Thanks for the help getting it over the line, @AndyScherzinger.

Always a pleasure to help @dan0xii ❤️
